internal/process/defunct_processes.go

```diff
@@ -1,59 +1,90 @@
 package process
 
 import (
-	"fmt"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	"os"
 	"strconv"
 	"strings"
 )
 
-// DefunctProcesses returns the PIDs of all the zombie processes in the node
-func DefunctProcesses() ([]int, error) {
-	directories, err := os.Open("/proc")
+// ProcessFS is the process file system.
+const ProcessFS = "/proc"
+
+// Stat represents status information of a process from /proc/[pid]/stat.
+type Stat struct {
+	// Fname is the filename of the executable.
+	Fname string
+
+	// State is the state of the process.
+	State string
+}
+
+// DefunctProcesses returns the number of zombie processes in the node.
+func DefunctProcesses() (int, error) {
+	directories, err := os.Open(ProcessFS)
 	if err != nil {
-		return nil, err
+		return -1, err
 	}
 	defer directories.Close()
 
 	names, err := directories.Readdirnames(-1)
 	if err != nil {
-		return nil, err
+		return -1, err
 	}
 
-	var pids []int
+	var defunctCount int
 	for _, name := range names {
-		// processes have numeric names. If the name does not start with a number, skip.
+		// Processes have numeric names. If the name does not start with a number, skip.
 		if name[0] < '0' || name[0] > '9' {
 			continue
 		}
 
-		pid, err := strconv.ParseInt(name, 10, 0)
+		// If the name cannot be parsed to an int, it is not numeric.
+		_, err := strconv.ParseInt(name, 10, 0)
 		if err != nil {
 			continue
 		}
 
-		stat, err := status(int(pid))
+		stat, err := status(name)
 		if err != nil {
+			logrus.Warnf("Failed to get the status of process with PID %s: %v", name, err)
 			continue
 		}
-		if stat == "Z" {
-			pids = append(pids, int(pid))
+		if stat.State == "Z" {
+			logrus.Warnf("Found defunct process with PID %s (%s)", name, stat.Fname)
+			defunctCount++
 		}
 	}
-	return pids, nil
+	return defunctCount, nil
 }
 
-// status returns the status of a process
-func status(pid int) (string, error) {
-	bytes, err := os.ReadFile("/proc/" + strconv.Itoa(pid) + "/stat")
+// status returns status information of a process as defined in /proc/[pid]/stat
+func status(pid string) (*Stat, error) {
+	bytes, err := os.ReadFile(ProcessFS + "/" + pid + "/stat")
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	data := string(bytes)
 
+	// Get the last index of field 2 in /proc/[pid]/stat.
+	// The last index of field 2 must contain a ')'.
```

Comment:
I would say something like

```go
// /proc/[PID]/stat format is described in proc(5). The second field is process name,
// enclosed in parentheses, and it can contain parentheses inside. No other fields
// can have parentheses, so we are looking for the last one.
```